theano/gof/optdb.py

```diff
@@ -306,19 +326,34 @@ def query(self, *tags, **kwtags):
 
         position_cutoff = kwtags.pop('position_cutoff',
                                      config.optdb.position_cutoff)
+        position_dict = self.__position__
+
         if len(tags) >= 1 and isinstance(tags[0], Query):
             # the call to super should have raise an error with a good message
             assert len(tags) == 1
             if getattr(tags[0], 'position_cutoff', None):
                 position_cutoff = tags[0].position_cutoff
 
-        opts = [o for o in opts if self.__position__[o.name] < position_cutoff]
+            # The Query instance might contain extra optimizations which need
+            # to be added the the sequence of optimizations (don't alter the
+            # original dictionary)
+            position_dict = position_dict.copy()
+            for extra_opt in tags[0].extra_optimizations:
+                # Give a name to the extra optimization
+                opt, position = extra_opt
+                opt.name = str(opt.__class__)
+
+                # Add the extra optimization to the optimization sequence
+                opts.add(opt)
+                position_dict[opt.name] = position
```

Comment:
As an optimization could you skip adding the opt if the position cutoff is smaller than the desired position.